File: javascript/algorithms/priority.js
const createArrivalTimeComparator = require("../comparators/arrivalTime");
const createPriorityComparator = require("../comparators/priority");

const priority = (processes, priorityDirection) => {
  let sortedProcesses = processes.sort(
    createPriorityComparator(
      createArrivalTimeComparator(),
      priorityDirection ? priorityDirection : "desc"
    )
  );

  let currentExecutionTime = 0;
  let totalWaitingTime = 0;
  for (let index = 0; index < sortedProcesses.length; index++) {
    let process = sortedProcesses[index];

    let executedAt = currentExecutionTime;
    currentExecutionTime += process.burstTime;
    let waitingTime = executedAt - process.arrivalTime;
    if (waitingTime < 0) {
      waitingTime = 0;
    }
    totalWaitingTime += waitingTime;
  }

  return totalWaitingTime / sortedProcesses.length;
};

module.exports = priority;

File: java/src/br/com/brunohpaiva/osprocessscheduling/algorithms/Priority.java
package br.com.brunohpaiva.osprocessscheduling.algorithms;

import br.com.brunohpaiva.osprocessscheduling.comparators.ArrivalTimeComparator;
import br.com.brunohpaiva.osprocessscheduling.comparators.PriorityComparator;
import br.com.brunohpaiva.osprocessscheduling.models.Process;

import java.util.Comparator;
import java.util.List;

public class Priority {

    public static Double calculate(List<Process> processes, Comparator<Process> tiebreakerComparator) {
        processes.sort(new PriorityComparator("desc", new ArrivalTimeComparator()));

        int currentExecutionTime = 0;
        double totalWaitingTime = 0.0;

        for (Process process : processes) {
            int executedAt = currentExecutionTime;
            currentExecutionTime += process.getBurstTime();
            int waitingTime = executedAt - process.getArrivalTime();
            if (waitingTime < 0)
                waitingTime = 0;
            totalWaitingTime += waitingTime;
        }

        return totalWaitingTime / processes.size();
    }

}
